src/util/compat.hpp

```diff
@@ -34,4 +34,27 @@ inline void compiler_barrier_if_xlc_leq(unsigned ver) {
 template <typename X>
 inline constexpr bool isinf(X x) { return std::isinf(x); }
 
+// Work around a bad inlining-related optimization with icpc 16.0.3 and -xMIC-AVX512,
+// by forcing a computation.
+
+template <typename X>
+inline void sink(const X& x) {
```

Comment:
I think the key here is the `volatile` keyword, which is just there to prevent compiler optimisations. The `buf` doesn't have any real meaning, it's an automatic variable, initialised but not used, which would have been otherwise optimised out. In fact, we could fool the compiler with just a single-char buffer avoiding the full copy of `x`:

```C++
char buf;
volatile char *bufptr = &buf;
const char* xptr = reinterpret_cast<const char*>(&x);
*bufptr = *xptr;
```

I tried this with Intel 16 on Tave and worked fine.